File: app/provider/layout_provider.py
from app.domain.layout import Layout
from app.repository.layout_repository import LayoutRepository


class LayoutProvider:
    
    _layout_repository: LayoutRepository
    
    def __init__(self):
        self._layout_repository = LayoutRepository()
    
    def add_layout(self, layout: Layout, updateOnDuplicate=False):    
        return self._layout_repository.add_layout(layout)
    
    def get_layout(self, key: str, value: str):
        return self._layout_repository.get_layout(key, value)
    
    def remove_layout(self, key: str, value):
        self._layout_repository.remove_layout(key, value)
    
    def get_all_layouts(self):
        return self._layout_repository.get_all_layouts()
        
    def update_layout(self, layout):
        return self._layout_repository.update_layout(layout)

File: kahldeck-dashboard/src/providers/LayoutProvider.ts
import LayoutRepositoryInterface from '../repositories/LayoutRepositoryInterface';
import { Layout } from '../domain/Layout';
import EelLayoutRepository from '../repositories/EelLayoutRepository';
import DummyLayoutRepository from '../repositories/dummies/DummyLayoutRepository';


export default class LayoutProvider {
    private static instance: LayoutProvider;
    private layoutRepository: LayoutRepositoryInterface;


    public constructor() {
        if( import.meta.env.PROD ) {
            this.layoutRepository = new EelLayoutRepository();
        } else {
            this.layoutRepository = new DummyLayoutRepository();
        }
        this.layoutRepository = new EelLayoutRepository();
    }

    static getInstance(): LayoutProvider {

        if (!LayoutProvider.instance) {
            LayoutProvider.instance = new LayoutProvider();
        }
        
        return LayoutProvider.instance;
    }

    async getLayouts(): Promise<Layout[]> {
        return this.layoutRepository.getLayouts();
    }

    async addLayout(layout: Layout): Promise<Layout> {
        await this.layoutRepository.addLayout(layout);
        return layout;
    }

    async removeLayout(uuid: string): Promise<void> {
        await this.layoutRepository.removeLayout(uuid);
    }
    
    async updateLayout(layout: Layout): Promise<Layout> {
        console.log('call updateLayout on provider');
        await this.layoutRepository.updateLayout(layout);
        return layout;
    }

    async dummy(text: string): Promise<string> {
        return this.layoutRepository.dummy(text);
    }

}